src/components/data/QueryAuth0.js

```diff
@@ -21,15 +21,35 @@ import { finishAuthenticationFlow } from 'console/utils/auth0';
 export default class QueryAuth0 extends React.PureComponent {
   static propTypes = {
     accessToken: PropTypes.string,
-    fetchUserProfile: PropTypes.func.isRequired,
+    userProfileReceived: PropTypes.func.isRequired,
     history: PropTypes.object.isRequired,
     loginFailed: PropTypes.func.isRequired,
     logUserIn: PropTypes.func.isRequired,
   };
 
+  // The motivation for using componentWILLMount instead of the more regular componentDidMount
+  // is because we don't worry about server-side rendering React AND there's a chance we
+  // can synchronously read from localStorage and update state such that we have the user
+  // profile *before* the AuthButton component is mounted. Thus, the the AuthButton might
+  // not have to render the "Log in" button at all since this method will dispatch the
+  // user profile before it renders.
   async componentWillMount() {
```

Comment:
I think this got overly complicated and can be simplified to:
```js
  async componentWillMount() {
    const { history, loginFailed, logUserIn, userProfileReceived } = this.props;
    let authResult;

    try {
      authResult = await finishAuthenticationFlow();
    } catch (err) {
      loginFailed(err);
    }

    if (!authResult) {
      authResult = JSON.parse(localStorage.getItem('authResult'));
    }

    if (authResult) {
      const { state } = authResult;
      logUserIn(authResult);
      userProfileReceived(authResult.idTokenpayload);

      if (state) {
        history.push(state);
      }
    }
  }
```

Reply:
I'm not too concerned about calling `userProfileReceived` when `sub` hasn't changed since this action has very low overhead.